### PR TITLE
Timeout updates

### PR DESCRIPTION
**Pull request by jmfee-usgs**
- Add connect/read timeouts for HTTPURLConnections, based on observed hanging threads:

```
"pool-1-thread-1" #15 prio=5 os_prio=0 cpu=857030.81ms elapsed=284571.09s tid=0x00007f92e049a000 nid=0x4f80 runnable  [0x00007f92b446b000]
   java.lang.Thread.State: RUNNABLE
        at java.net.SocketInputStream.socketRead0(java.base@11.0.9.1/Native Method)
        at java.net.SocketInputStream.socketRead(java.base@11.0.9.1/SocketInputStream.java:115)
        at java.net.SocketInputStream.read(java.base@11.0.9.1/SocketInputStream.java:168)
        at java.net.SocketInputStream.read(java.base@11.0.9.1/SocketInputStream.java:140)
        at sun.security.ssl.SSLSocketInputRecord.read(java.base@11.0.9.1/SSLSocketInputRecord.java:476)
        at sun.security.ssl.SSLSocketInputRecord.readHeader(java.base@11.0.9.1/SSLSocketInputRecord.java:470)
        at sun.security.ssl.SSLSocketInputRecord.bytesInCompletePacket(java.base@11.0.9.1/SSLSocketInputRecord.java:70)
        at sun.security.ssl.SSLSocketImpl.readApplicationRecord(java.base@11.0.9.1/SSLSocketImpl.java:1354)
        at sun.security.ssl.SSLSocketImpl$AppInputStream.read(java.base@11.0.9.1/SSLSocketImpl.java:963)
        at java.io.BufferedInputStream.fill(java.base@11.0.9.1/BufferedInputStream.java:252)
        at java.io.BufferedInputStream.read1(java.base@11.0.9.1/BufferedInputStream.java:292)
        at java.io.BufferedInputStream.read(java.base@11.0.9.1/BufferedInputStream.java:351)
        - locked <0x00000000c444a9c8> (a java.io.BufferedInputStream)
        at sun.net.www.http.HttpClient.parseHTTPHeader(java.base@11.0.9.1/HttpClient.java:754)
        at sun.net.www.http.HttpClient.parseHTTP(java.base@11.0.9.1/HttpClient.java:689)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(java.base@11.0.9.1/HttpURLConnection.java:1615)
        - locked <0x00000000c444aa20> (a sun.net.www.protocol.https.DelegateHttpsURLConnection)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(java.base@11.0.9.1/HttpURLConnection.java:1520)
        - locked <0x00000000c444aa20> (a sun.net.www.protocol.https.DelegateHttpsURLConnection)
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(java.base@11.0.9.1/HttpsURLConnectionImpl.java:250)
        - locked <0x00000000c444ab90> (a sun.net.www.protocol.https.HttpsURLConnectionImpl)
        at gov.usgs.earthquake.aws.HttpResponse.<init>(HttpResponse.java:28)
        at gov.usgs.earthquake.aws.AwsProductSender.postProductJson(AwsProductSender.java:195)
        at gov.usgs.earthquake.aws.AwsProductSender.getUploadUrls(AwsProductSender.java:154)
        at gov.usgs.earthquake.aws.AwsProductSender.sendProduct(AwsProductSender.java:113)
        at gov.usgs.earthquake.distribution.RelayProductListener.onProduct(RelayProductListener.java:54)
        at gov.usgs.earthquake.distribution.DefaultNotificationListener.onNotification(DefaultNotificationListener.java:130)
        at gov.usgs.earthquake.distribution.NotificationListenerCallable.call(NotificationListenerCallable.java:34)
        at gov.usgs.earthquake.distribution.NotificationListenerCallable.call(NotificationListenerCallable.java:10)
        at gov.usgs.util.ExecutorTask.run(ExecutorTask.java:183)
        at java.util.concurrent.Executors$RunnableAdapter.call(java.base@11.0.9.1/Executors.java:515)
        at java.util.concurrent.FutureTask.run(java.base@11.0.9.1/FutureTask.java:264)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@11.0.9.1/ThreadPoolExecutor.java:1128)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@11.0.9.1/ThreadPoolExecutor.java:628)
        at java.lang.Thread.run(java.base@11.0.9.1/Thread.java:829)
```

- cancel Timers, and use Futures for timeouts to reduce thread creation.  Many references showed up in heap dump.

### CODE REVIEW

**Review by emartinez-usgs — CHANGES_REQUESTED** [review]
Just a small change regarding duplicate `done` references. This was an existing issue and hasn't caused problems yet, so fine to push this fix off to a future enhancement.

File: src/main/java/gov/usgs/util/ExecutorTask.java [hunk 1]
```diff
@@ -223,7 +223,7 @@ public void run() {
 	 * Called when task is completed, either successfully, or unsuccessfully and
 	 * has no more tries
 	 */
-	private void done() {
+	protected void done() {
```
[the remark this answers is not in the document]
Comment (jmfee-usgs) [reply src/main/java/gov/usgs/util/ExecutorTask.java]: Renamed `done()` method to `setDone()` to make this clearer

File: src/main/java/gov/usgs/earthquake/aws/AwsProductSender.java [hunk 2]
```diff
@@ -116,7 +116,24 @@ public void sendProduct(final Product product) throws Exception {
...
+        try {
+          uploadProduct = getUploadUrls(json);
+        } catch (HttpException e) {
+          HttpURLConnection connection = e.response.connection;
+          // check for server error
+          if (connection.getResponseCode() >= 500) {
+            LOGGER.log(Level.FINE,
+                "[" + getName() + "] get upload urls exception, trying again", e);
+            // try again after random back off (1-5 s)
+            Thread.sleep(1000 + Math.round(4000 * Math.random()));
```
Comment (emartinez-usgs) [inline src/main/java/gov/usgs/earthquake/aws/AwsProductSender.java:129]: This is probably fine, but AWS does recommend and exponential backoff. Maybe 10 * (10^retry) ms ? Not sure the retry number is tracked?

**Review by emartinez-usgs — APPROVED** [review]
There are some pretty big changes in here. Things _look_ okay, but hard to track everything. Would you want to bump to 2.7.1 since some changes occurred after 2.7.0 or since it wasn't released, it's fine to just stay at 2.7.0?

File: src/main/java/gov/usgs/earthquake/aws/AwsProductSender.java (hunk 2 above)
Comment (jmfee-usgs) [reply src/main/java/gov/usgs/earthquake/aws/AwsProductSender.java:129]: Yeah, retries are tracked outside the listener...  Those retries jump to the end of the line, while this is an additional attempt inline.

**Pull request merged**